src/wasm2asm.h

```diff
@@ -221,171 +223,16 @@ class Wasm2AsmBuilder {
-         op == RotLInt64 || op == RotRInt64);
-  Builder b(allocator);
-  // left rotate is:
-  // (((((~0) >>> k) & x) << k) | ((((~0) << (w - k)) & x) >>> (w - k)))
-  // where k is shift modulo w. reverse shifts for right rotate
-  bool is32Bit = (op == RotLInt32 || op == RotRInt32);
-  bool isLRot  = (op == RotLInt32 || op == RotLInt64);
-  static Name names[2][2] = {{Name(WASM_ROTR64), Name(WASM_ROTR32)},
-                             {Name(WASM_ROTL64), Name(WASM_ROTL32)}};
-  static BinaryOp shifters[2][2] = {{ShrUInt64, ShrUInt32},
-                                    {ShlInt64, ShlInt32}};
-  Name funcName = names[isLRot][is32Bit];
-  BinaryOp lshift = shifters[isLRot][is32Bit];
-  BinaryOp rshift = shifters[!isLRot][is32Bit];
-  BinaryOp orOp    = is32Bit ? OrInt32  : OrInt64;
-  BinaryOp andOp   = is32Bit ? AndInt32 : AndInt64;
-  BinaryOp subOp   = is32Bit ? SubInt32 : SubInt64;
-  Type argType = is32Bit ? i32      : i64;
-  Literal widthMask =
-      is32Bit ? Literal(int32_t(32 - 1)) : Literal(int64_t(64 - 1));
-  Literal width =
-      is32Bit ? Literal(int32_t(32)) : Literal(int64_t(64));
-  auto shiftVal = [&]() {
-    return b.makeBinary(
-      andOp,
-      b.makeGetLocal(1, argType),
-      b.makeConst(widthMask)
-    );
-  };
-  auto widthSub = [&]() {
-    return b.makeBinary(subOp, b.makeConst(width), shiftVal());
-  };
-  auto fullMask = [&]() {
-    return b.makeConst(is32Bit ? Literal(~int32_t(0)) : Literal(~int64_t(0)));
-  };
-  Binary* maskRShift = b.makeBinary(rshift, fullMask(), shiftVal());
-  Binary* lowMask = b.makeBinary(andOp, maskRShift, b.makeGetLocal(0, argType));
-  Binary* lowShift = b.makeBinary(lshift, lowMask, shiftVal());
-  Binary* maskLShift = b.makeBinary(lshift, fullMask(), widthSub());
-  Binary* highMask =
-      b.makeBinary(andOp, maskLShift, b.makeGetLocal(0, argType));
-  Binary* highShift = b.makeBinary(rshift, highMask, widthSub());
-  Binary* body = b.makeBinary(orOp, lowShift, highShift);
-  return b.makeFunction(
-    funcName,
-    std::vector<NameType>{NameType("x", argType),
-          NameType("k", argType)},
-    argType,
-    std::vector<NameType>{},
-    body
-  );
-}
-
-void Wasm2AsmBuilder::addWasmCompatibilityFuncs(Module* wasm) {
-  wasm->addFunction(makeCtzFunc(wasm->allocator, CtzInt32));
-  wasm->addFunction(makePopcntFunc(wasm->allocator, PopcntInt32));
-  wasm->addFunction(makeRotFunc(wasm->allocator, RotLInt32));
-  wasm->addFunction(makeRotFunc(wasm->allocator, RotRInt32));
-}
-
 Ref Wasm2AsmBuilder::processWasm(Module* wasm) {
-  addWasmCompatibilityFuncs(wasm);
   PassRunner runner(wasm);
   runner.add<AutoDrop>();
-  runner.add("remove-copysign"); // must be before i64-to-i32
+  runner.add("remove-non-asm-ops"); // must be before i64-to-i32
+  // TODO: it is a bug that `flatten` is required before the `i64-to-i32`
+  // lowering pass, but it's a hack that works for now. See #1480 for more
+  // details.
+  runner.add("flatten");
```

Comment:
instead of documenting this as a bug, we can document the i64 lowering pass as requiring flat IR. seems reasonable.

slightly sad we need to flatten twice, though, in theory that pass could keep the IR flat to avoid it, but whatever.

Reply:
Ah yeah I just found that this fixed a bug or two in some of the spec tests so I threw it in, but I'm not acutally certain whether it's necessary aftewards as well, I just assumed.